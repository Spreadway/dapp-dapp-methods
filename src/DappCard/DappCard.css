.picture {
  display: block !important;
  clear: right;
  height: 60px;
  width: 60px;
  border-radius: 50%;
  margin-bottom: 20px;
}

.edit {
  margin-bottom: -15px !important;
}

.list {
  max-height: 250px;
  overflow-y: scroll;
}

.item {
  cursor: pointer;
}

.item:hover {
  opacity: 0.8;
}
