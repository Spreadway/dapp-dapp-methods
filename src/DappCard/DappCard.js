// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { toJS } from 'mobx';
import { Accordion, Button, Card, Image, List } from 'semantic-ui-react';
import { FormattedMessage } from 'react-intl';
import flatten from 'lodash/flatten';

import { getPermissionId } from '../store';
import styles from './DappCard.css';

class DappCard extends PureComponent {
  static propTypes = {
    dapp: PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      image: PropTypes.string
    }).isRequired,
    editingMode: PropTypes.bool,
    methodGroups: PropTypes.object.isRequired,
    onEdit: PropTypes.func.isRequired,
    onToggle: PropTypes.func.isRequired,
    permissions: PropTypes.object.isRequired
  };

  renderEditingMode = () => {
    const { dapp, methodGroups, permissions, onToggle } = this.props;

    const panels = Object.keys(methodGroups).map(group => ({
      title: {
        key: `${dapp.id}-${group}-editing-title`,
        active: true,
        content: <FormattedMessage {...messages[group]} />
      },
      content: {
        key: `${dapp.id}-${group}-editing-content`,
        active: true,
        content: (
          <List className={styles.editList}>
            {methodGroups[group].methods.map(method => (
              <List.Item
                key={method}
                onClick={() => onToggle(method, dapp.id)}
                className={styles.editListItem}
              >
                <List.Icon
                  name={
                    permissions[getPermissionId(method, dapp.id)]
                      ? 'checkmark box'
                      : 'square outline'
                  }
                />
                {method}
              </List.Item>
            ))}
          </List>
        )
      }
    }));

    return <Accordion panels={panels} />;
  };

  renderViewMode = () => {
    const { methodGroups, permissions, dapp } = this.props;

    if (
      !methodGroups ||
      !flatten(
        Object.values(toJS(methodGroups)).map(({ methods }) => methods)
      ).some(method => permissions[getPermissionId(method, dapp.id)])
    ) {
      return (
        <div className={styles.noAllowedMethods}>
          <FormattedMessage
            id="dapps.methods.noAllowedMethods"
            defaultMessage="No allowed methods"
          />
        </div>
      );
    }

    const panels = Object.keys(methodGroups)
      .filter(group =>
        methodGroups[group].methods.some(
          method => permissions[getPermissionId(method, dapp.id)]
        )
      )
      .map(group => ({
        title: {
          key: `${dapp.id}-${group}-title`,
          active: true,
          content: <FormattedMessage {...messages[group]} />
        },
        content: {
          key: `${dapp.id}-${group}-content`,
          content: (
            <List bulleted className={styles.list}>
              {methodGroups[group].methods.map(
                method =>
                  permissions[getPermissionId(method, dapp.id)] && (
                    <List.Item key={method}>{method}</List.Item>
                  )
              )}
            </List>
          )
        }
      }));

    return <Accordion panels={panels} />;
  };

  render() {
    const { dapp, editingMode, onEdit } = this.props;

    return (
      <Card>
        <Card.Content>
          <Button
            floated="right"
            basic
            size="mini"
            icon={editingMode ? 'remove' : 'edit'}
            className={styles.editButton}
            onClick={onEdit}
          />
          <Image src={dapp.image} className={styles.picture} centered />
          <Card.Header>{dapp.name}</Card.Header>
          <Card.Meta>{dapp.description}</Card.Meta>
          <Card.Description>
            {editingMode ? this.renderEditingMode() : this.renderViewMode()}
          </Card.Description>
        </Card.Content>
      </Card>
    );
  }
}

export default DappCard;

const messages = {
  shell: {
    id: 'dapp.methods.shell',
    description: 'Explanation of the shell methodGroup permission',
    defaultMessage: 'Can get access to the Shell'
  },
  accountsView: {
    id: 'dapp.methods.accountsView',
    description: 'Explanation of the accountsView methodGroup permission',
    defaultMessage: 'Can view your Parity accounts'
  },
  accountsCreate: {
    id: 'dapp.methods.accountsCreate',
    description: 'Explanation of the accountsCreate methodGroup permission',
    defaultMessage: 'Can create new Parity accounts'
  },
  accountsEdit: {
    id: 'dapp.methods.accountsEdit',
    description: 'Explanation of the accountsEdit methodGroup permission',
    defaultMessage: 'Can edit your Parity accounts'
  },
  upgrade: {
    id: 'dapp.methods.upgrade',
    description: 'Explanation of the upgrade methodGroup permission',
    defaultMessage: 'Can upgrade Parity'
  },
  vaults: {
    id: 'dapp.methods.vaults',
    description: 'Explanation of the vaults methodGroup permission',
    defaultMessage: 'Can get access to your vaults'
  },
  other: {
    id: 'dapp.methods.other',
    description: 'Explanation of the other methodGroup permission',
    defaultMessage: 'Other permissions'
  }
};
