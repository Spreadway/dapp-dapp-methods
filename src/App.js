// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import { observer } from 'mobx-react';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

import Card from 'semantic-ui-react/dist/commonjs/views/Card';
import ActionBar from '@parity/ui/lib/Actionbar/actionbar';

import DappCard from './DappCard';
import Store from './store';
import styles from './App.css';

class App extends Component {
  state = {
    selectedDapp: null
  };

  store = new Store(this.context.api);

  handleSelectDapp = id => {
    if (this.state.selectedDapp === id) {
      this.setState({ selectedDapp: null });
    } else {
      this.setState({ selectedDapp: id });
    }
  };

  render() {
    return (
      <div className={styles.layout}>
        <ActionBar
          title={
            <FormattedMessage
              id="dapps.methods.title"
              defaultMessage="Allowed methods"
            />
          }
        />
        <Card.Group stackable className={styles.cardGroup}>
          {this.store.apps.map((dapp, index) => (
            <DappCard
              key={index}
              editingMode={dapp.id === this.state.selectedDapp}
              dapp={dapp}
              methodGroups={this.store.methodGroups}
              permissions={this.store.permissions}
              onEdit={() => this.handleSelectDapp(dapp.id)}
              onToggle={this.store.toggleAppPermission}
            />
          ))}
        </Card.Group>
      </div>
    );
  }
}

export default observer(App);

App.contextTypes = {
  api: PropTypes.object.isRequired
};
