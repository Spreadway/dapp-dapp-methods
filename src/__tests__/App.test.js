// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React from 'react';
import { shallowToJson } from 'enzyme-to-json';

import { shallowWithIntl } from '../setupTests';
import App from '../App';
import DappCard from '../DappCard';

const props = {
  store: {
    apps: [{ id: '123', name: '123' }, { id: '456', name: '456' }],
    methodGroups: { shell: { methods: ['foo', 'bar'] } },
    permissions: { 'foo:123': true },
    toggleAppPermission: () => {}
  }
};

test('should render correctly', () => {
  const component = shallowWithIntl(<App {...props} />);

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should select a dapp when we click on its edit button', () => {
  const component = shallowWithIntl(<App {...props} />);

  component
    .find(DappCard)
    .last()
    .props()
    .onEdit();

  expect(component.state().selectedDapp).toEqual('456');
});

test('should render correctly when a dapp is selected', () => {
  const component = shallowWithIntl(<App {...props} />);
  component.setState({ selectedDapp: '456' });

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should deselect a dapp when we click again on its edit button', () => {
  const component = shallowWithIntl(<App {...props} />);
  component.setState({ selectedDapp: '456' });

  component
    .find(DappCard)
    .last()
    .props()
    .onEdit();

  expect(component.state().selectedDapp).toEqual(null);
});

test('should call onToggleAppPermission when we toggle permission', () => {
  const toggleAppPermission = jest.fn();
  const component = shallowWithIntl(
    <App store={{ ...props.store, toggleAppPermission }} />
  );

  component
    .find(DappCard)
    .last()
    .props()
    .onToggle();

  expect(toggleAppPermission).toHaveBeenCalled();
});
