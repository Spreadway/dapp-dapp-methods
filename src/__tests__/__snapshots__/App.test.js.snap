// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <Actionbar
    title={
      <FormattedMessage
        defaultMessage="Allowed methods"
        id="dapps.methods.title"
        values={Object {}}
      />
    }
  />
  <CardGroup
    stackable={true}
  >
    <DappCard
      dapp={
        Object {
          "id": "123",
          "name": "123",
        }
      }
      editingMode={false}
      key="123"
      methodGroups={
        Object {
          "shell": Object {
            "methods": Array [
              "foo",
              "bar",
            ],
          },
        }
      }
      onEdit={[Function]}
      onToggle={[Function]}
      permissions={
        Object {
          "foo:123": true,
        }
      }
    />
    <DappCard
      dapp={
        Object {
          "id": "456",
          "name": "456",
        }
      }
      editingMode={false}
      key="456"
      methodGroups={
        Object {
          "shell": Object {
            "methods": Array [
              "foo",
              "bar",
            ],
          },
        }
      }
      onEdit={[Function]}
      onToggle={[Function]}
      permissions={
        Object {
          "foo:123": true,
        }
      }
    />
  </CardGroup>
</div>
`;

exports[`should render correctly when a dapp is selected 1`] = `
<div>
  <Actionbar
    title={
      <FormattedMessage
        defaultMessage="Allowed methods"
        id="dapps.methods.title"
        values={Object {}}
      />
    }
  />
  <CardGroup
    stackable={true}
  >
    <DappCard
      dapp={
        Object {
          "id": "123",
          "name": "123",
        }
      }
      editingMode={false}
      key="123"
      methodGroups={
        Object {
          "shell": Object {
            "methods": Array [
              "foo",
              "bar",
            ],
          },
        }
      }
      onEdit={[Function]}
      onToggle={[Function]}
      permissions={
        Object {
          "foo:123": true,
        }
      }
    />
    <DappCard
      dapp={
        Object {
          "id": "456",
          "name": "456",
        }
      }
      editingMode={true}
      key="456"
      methodGroups={
        Object {
          "shell": Object {
            "methods": Array [
              "foo",
              "bar",
            ],
          },
        }
      }
      onEdit={[Function]}
      onToggle={[Function]}
      permissions={
        Object {
          "foo:123": true,
        }
      }
    />
  </CardGroup>
</div>
`;
